include/zephyr/device.h

```diff
@@ -962,6 +962,35 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 			Z_DEVICE_INIT(name, pm, data, config, api, state,      \
 				      handles)
 
+/**
+ * @brief Obtain wrapper function name for device init.
+ *
+ * @param dev_id Device identifier.
+ */
+#define Z_DEVICE_INIT_WRAPPER_FN_NAME(dev_id) _CONCAT(__device_init_, dev_id)
+
+/**
+ * @brief Define wrapper function for device initialization.
+ *
+ * @param dev_id Device identifier.
+ * @param init_fn Init function.
+ */
+#define Z_DEVICE_INIT_WRAPPER_FN_DEFINE(dev_id, init_fn)                       \
+	static inline int Z_DEVICE_INIT_WRAPPER_FN_NAME(dev_id)(void *data)    \
```

Comment:
Isn't this going to result in an instance of this function for every device in the build?

Reply:
It  is indeed, this should be moved to `device.c` as a singleton instead.

Reply:
That can't be done without storing the `init_fn` in `struct device`, the singleton would have no way to retrieve the function otherwise